### PR TITLE
Docs/feature rework

### PR DESCRIPTION
**Pull request by Atahanak**
Feature extraction tutorial and how to add a feature pages are updated in accordance with the latest updates.

### CODE REVIEW

File: docs/pages/how_to_guides/how_add_feature.md [hunk 1]
```diff
@@ -52,13 +54,13 @@ template <typename IDType, typename NNZType, typename ValueType, typename Featur
 class Feature : FeaturePreprocessType<IDType, NNZType, ValueType, FeatureType> {
```
Comment (AmroAlJundi) [inline docs/pages/how_to_guides/how_add_feature.md:54]: FeaturePreprocessType templates should be changed here.

File: docs/pages/how_to_guides/how_add_feature.md [hunk 2]
```diff
@@ -52,13 +54,13 @@ template <typename IDType, typename NNZType, typename ValueType, typename Featur
...
 	// ...
 	Feature(float alpha, float beta){
-		this->params_ = shared_ptr<FeatureParams>(new FeatureParams{alpha, beta});
+		this->params_ = std::make_shared<FeatureParams>(new FeatureParams{alpha, beta});
```
Comment (AmroAlJundi) [inline docs/pages/how_to_guides/how_add_feature.md:57]: As parameters to make shared you only need the parameters to the constructor of `FeatureParams`, so:
`std::make_shared<FeatureParams>(alpha, beta);`

File: docs/pages/how_to_guides/how_add_feature.md [hunk 3]
```diff
@@ -123,24 +125,38 @@ class Feature : FeaturePreprocessType<IDType, NNZType, ValueType, FeatureType> {
...
-```cpp
-template class Feature<unsigned int, unsigned int, float, double>;
+1. Move all the implementations from the header file (`src/sparsebase/preprocess/preprocess.h`) to the implementation file (`src/sparsebase/preprocess/preprocess.cc`).
+2. Add your class to the list of classes that will be explicitly instantiated by the python script `src/generate_explicit_instantiations.py`.
+
+Step two is much simpler than it sounds. To the file `src/class_instantiation_list.json`, add a single entry containing the `Feature` class definition and the name of the file to which the instantiations are to be printed. In the case of `Feature`, add the following entry to the aformentioned file:
+
+```json
+{
+  "template": "class Feature<$id_type, $nnz_type, $value_type, $feature_type>",
```
Comment (AmroAlJundi) [inline docs/pages/how_to_guides/how_add_feature.md:146]: The last parameter should be `$float_type`.

**Review by AmroAlJundi — CHANGES_REQUESTED** [review]
Left a few comments to be addressed.

File: docs/pages/how_to_guides/how_add_feature.md (hunk 3 above)
Comment (Atahanak) [reply docs/pages/how_to_guides/how_add_feature.md:146]: But the feature types does not have to be a float.

File: docs/pages/how_to_guides/how_add_feature.md [hunk 4]
```diff
@@ -2,29 +2,29 @@
...
 ## Overview
 
-Adding new features to SparseBase is very simple. It consists of six steps:
+Adding new features to SparseBase is simple. It consists of six steps:
 
-1. create a new class for your feature. It must inherit from the base `FeaturePreprocessType` class.
-2. `FeaturePreprocessType` implements the `ExtractableType` interface thus all virtual functions defined by `ExtractableType` must be implemented. `ExtractableType` provides a seamless interface for feature extraction.
+1. Create a new class for your feature. It must inherit from the base `FeaturePreprocessType` class.
+2. `FeaturePreprocessType` implements the `ExtractableType` interface thus all virtual functions defined by `ExtractableType` must be implemented. `ExtractableType` provides a template for all the functionality needed for feature extraction.
 3. Inside the class, create a new struct that will contain the parameters needed by your feature, and initialize it in the constructor. Note: This step is only needed if your feature requires parameters to work with.
```
[the remark this answers is not in the document]
Comment (Atahanak) [reply docs/pages/how_to_guides/how_add_feature.md:9]: So if the feature does not require additional parameters the `params_` variable can be initialized with `PreprocessParams`. That's what I meant by this note. I will add a note to the following step clarifying this.

File: docs/pages/how_to_guides/how_add_feature.md [hunk 5]
```diff
@@ -31,7 +31,7 @@ class Feature : FeaturePreprocessType<FeatureType> {
 
 ### 2. Create a struct containing the parameters you need, and initialize them in the constructor
 
-Inside the class, create a new struct inheriting from `PreprocessParams`. Its members will be whichever parameters that your feature will require. We will call this struct `FeatureParams`. We add `alpha` and `beta` to it.
+Inside the class, create a new struct inheriting from `PreprocessParams`. Its members will be whichever parameters that your feature will require. We will call this struct `FeatureParams`. We add `alpha` and `beta` to it. If your feature do not require additional parameters you can skip this step.
```
Comment (AmroAlJundi) [inline docs/pages/how_to_guides/how_add_feature.md:34]: Actually it should be outside the class :)

File: docs/pages/how_to_guides/how_add_feature.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (AmroAlJundi) [reply docs/pages/how_to_guides/how_add_feature.md:9]: I see, makes sense.

File: docs/pages/how_to_guides/how_add_feature.md (hunk 3 above)
Comment (AmroAlJundi) [reply docs/pages/how_to_guides/how_add_feature.md:147]: We only have $id_type, $nnz_type, $value_type, and $float_type as tempalated variable lists provided by CMake, so we need to only use those in the template JSON.

**Review by AmroAlJundi — APPROVED** [review]

**Review by AmroAlJundi — APPROVED** [review]

**Pull request merged**